src/lib/components/DiagramReducer/diagram-reducer.tsx

```diff
@@ -0,0 +1,253 @@
+import dayjs, { Dayjs } from "dayjs";
+
+import { FlowDiagram, DiagramConfig } from "../../types/diagram";
+import { Diagram, DiagramDrawer } from "../../utils/diagram-drawer-advanced";
+
+type ActionMap<M extends { [index: string]: { [key: string]: string | Dayjs | number | FlowDiagram[] } }> = {
+    [Key in keyof M]: M[Key] extends undefined
+        ? {
+              type: Key;
+          }
+        : {
+              type: Key;
+              payload: M[Key];
+          };
+};
+
+export enum DiagramActionTypes {
+    MoveDown = "MOVE_DOWN",
+    MoveUpToNode = "MOVE_UP_TO_NODE",
+    ChangeDate = "CHANGE_DATE",
+    ChangeDiagram = "CHANGE_DIAGRAM",
+}
+
+type DiagramReducerStateType = {
+    fixed: {
+        flowDiagrams: FlowDiagram[];
+        diagramConfig: DiagramConfig;
+        globalStartDate: Dayjs;
+        globalEndDate: Dayjs;
+    };
+    currentPath: PathElement[];
+    currentDate: Dayjs;
+    currentDiagram?: Diagram;
+};
+
+type Payload = {
+    [DiagramActionTypes.MoveDown]: {
+        id: string;
+    };
+    [DiagramActionTypes.MoveUpToNode]: {
+        nodeId: string;
+    };
+    [DiagramActionTypes.ChangeDate]: {
+        date: Dayjs;
+    };
+    [DiagramActionTypes.ChangeDiagram]: {
+        diagram: FlowDiagram[];
+    };
+};
+
+type Actions = ActionMap<Payload>[keyof ActionMap<Payload>];
```

Comment:
Could you try to explain this, or link an explanation? What's the result of `ActionMap<Payload>` and what is happening when you write `[keyof ActionMap<Payload>]`?

Reply:
Sure. Let's start with the definition of `ActionMap`.

`type ActionMap<M extends { [index: string]: { [key: string]: string | Dayjs | number | FlowDiagram[] } }>`

- `ActionMap` is a type that is supposed to map action types (`MOVE_DOWN`, `CHANGE_DIAGRAM` etc. to the payload data).
- It uses a template `M` that must extend an object structure with 2 dimensions:
```
[key of type string]: { 
    [key of type string]: string | Dayjs | number | FlowDiagram[],
    ...
},
...
```
- In its type definition it defines a two dimensional object that takes the keys of the first dimension of `M` and checks if a second dimension is defined (`M[Key] extends undefined ?`). If it is not defined, the resulting object only has `type` as a second dimension attribute, else it also gets a payload (`M[Key]`).

Let's apply an example template to `ActionMap` for better understanding what is going on:

```
type actionTemplate = {
    actionWithPayload: {
        id: string;
    };
    actionWithoutPayload: {};
}; // actionTemplate must match the type of the template M defined in ActionMap!
        
ActionMap<actionTemplate> = { 
    actionWithPayload: {
        type: actionWithPayload;
        payload: {
            id: string;
        };
    };
    actionWithoutPayload: {
        type: actionWithoutPayload;
    };
};
```

The goal is to take an object of type`M` and make it a union of one dimensional objects where each element (action) definitely holds a `type` and optionally a `payload` attribute. In order to achieve this, we have to extract the second dimension of `ActionMap`. This is what`keyof` is used for, which creates a union of all keys in the given map. In our example above it would yield:

```
keyof ActionMap<actionTemplate> = "actionWithPayload" | "actionWithoutPayload";
```

If we now apply this logic to the aforedefined `ActionMap`, we achieve our goal:

```
ActionMap<actionTemplate>[keyof ActionMap<actionTemplate>] = {
    type: "actionWithPayload";
    payload: {
        id: string;
    };
} | { 
    type: "actionWithoutPayload";
};
```

Now we have a type that allows any of the actions defined in our template `M` and consists of an object with one or two attributes (`type` and `payload`).

For further reading, I recommend taking a look at https://mariusschulz.com/blog/conditional-types-in-typescript

Reply:
Nice explanation, thanks!